### PR TITLE
feat: [CDS-40802]: Implemented AWS cloud provider migration.

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/38027)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 1 Issues 
* Code issue 1 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/38027)

**Comment by GoelPratyush99** [comment]
MessageMetadata: trigger messagecheck

File: 100-migrator/src/main/java/io/harness/ngmigration/service/entity/ConnectorMigrationService.java [hunk 1]
```diff
@@ -114,10 +114,15 @@ public DiscoveryNode discover(NGMigrationEntity entity) {
...
                                      .build();
     Set<CgEntityId> children = new HashSet<>();
-    String secret = ConnectorFactory.getConnector(settingAttribute).getSecretId(settingAttribute);
+    BaseConnector baseConnector = ConnectorFactory.getConnector(settingAttribute);
+    String secret = baseConnector.getSecretId(settingAttribute);
     if (StringUtils.isNotBlank(secret)) {
       children.add(CgEntityId.builder().id(secret).type(NGMigrationEntityType.SECRET).build());
     }
+    String accessId = ConnectorFactory.getAccessId(baseConnector, settingAttribute);
+    if (StringUtils.isNotBlank(accessId)) {
```
Comment (puthrayaharness) [inline 100-migrator/src/main/java/io/harness/ngmigration/service/entity/ConnectorMigrationService.java:123]: We should not be doing this. If the connector has multiple secrets then we need to support fetching multiple secrets.

File: 100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+  @Override
+  public ConnectorConfigDTO getConfigDTO(
+      SettingAttribute settingAttribute, Set<CgEntityId> childEntities, Map<CgEntityId, NGYamlFile> migratedEntities) {
+    AwsConfig clusterConfig = (AwsConfig) settingAttribute.getValue();
+    AwsConnectorDTOBuilder builder = builder().delegateSelectors(
+        clusterConfig.getTag() != null ? new HashSet<>(Arrays.asList(clusterConfig.getTag())) : null);
+    AwsCredentialDTO awsCredentialDTO = null;
+
+    if (clusterConfig.isUseEc2IamCredentials()) {
```
Comment (puthrayaharness) [inline 100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java:59]: Please create a separate function for each of the modes inside AWSConnectorImpl. Then just call the function here. The code will be more readable that way

File: 100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                            .crossAccountRoleArn(clusterConfig.getCrossAccountAttributes().getCrossAccountRoleArn())
+                            .externalId(clusterConfig.getCrossAccountAttributes().getExternalId())
+                            .build())
+              .config(AwsIRSASpecDTO.builder().build())
+              .testRegion(clusterConfig.getDefaultRegion())
+              .build();
+    }
+
+    else {
+      NGYamlFile secretRef = migratedEntities.get(
```
Comment (puthrayaharness) [inline 100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java:90]: Please use MigratorUtils.getSecretRef.

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

**Review by vikyathharekal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38027)* status: 3 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java` line 56 at r2](https://reviewable.io/reviews/harness/harness-core/38027#-ND1Raw79eL0tCS_6Z9x:-ND1Raw79eL0tCS_6Z9y:b-o8pw0q) ([raw file](https://github.com/harness/harness-core/blob/3e456575be87ecc61a0b7677e1c7419fd936fc7a/100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java#L56)):*
> ```Java
>     AwsConfig clusterConfig = (AwsConfig) settingAttribute.getValue();
>     AwsConnectorDTOBuilder builder = builder().delegateSelectors(
>         clusterConfig.getTag() != null ? new HashSet<>(Arrays.asList(clusterConfig.getTag())) : null);
> ```

Collections.singleton(clusterConfig.getTag()) instead?



_Code quote:_
```Java
new HashSet<>(Arrays.asList(clusterConfig.getTag())
```
___
*[`100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java` line 57 at r2](https://reviewable.io/reviews/harness/harness-core/38027#-ND1QwTFAYvksanLdJ8Y:-ND1QwTFAYvksanLdJ8Z:b2fpivn) ([raw file](https://github.com/harness/harness-core/blob/3e456575be87ecc61a0b7677e1c7419fd936fc7a/100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java#L57)):*
> ```Java
>     AwsConnectorDTOBuilder builder = builder().delegateSelectors(
>         clusterConfig.getTag() != null ? new HashSet<>(Arrays.asList(clusterConfig.getTag())) : null);
>     AwsCredentialDTO awsCredentialDTO = null;
> ```

nit: null initialization might not be needed.
___
*[`100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java` line 65 at r2](https://reviewable.io/reviews/harness/harness-core/38027#-ND1QfxrDx7WoQwyhFgq:-ND1QfxrDx7WoQwyhFgr:bpy2sth) ([raw file](https://github.com/harness/harness-core/blob/3e456575be87ecc61a0b7677e1c7419fd936fc7a/100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java#L65)):*
> ```Java
>               .crossAccountAccess(!clusterConfig.isAssumeCrossAccountRole()
>                       ? null
>                       : CrossAccountAccessDTO.builder()
> ```

We need the same CrossAccountAccessDTO for every mode. Can we build it outside and use it?



_Code quote:_
```Java
CrossAccountAccessDTO
```
___
*[`100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java` line 90 at r2](https://reviewable.io/reviews/harness/harness-core/38027#-ND12FGftubx7xo7taG_-r2-90:-ND1RHT2-qeWkn-33ejv:bas9r49) ([raw file](https://github.com/harness/harness-core/blob/3e456575be87ecc61a0b7677e1c7419fd936fc7a/100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java#L90)):*
<details><summary><i>Previously, puthrayaharness (Deepak Puthraya) wrote…</i></summary><blockquote>

Please use MigratorUtils.getSecretRef.
</blockquote></details>

@goelpratyush99 We could use this in Azure and HTTP helm as well.

**Review by GoelPratyush99 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38027)* status: 3 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java` line 56 at r2](https://reviewable.io/reviews/harness/harness-core/38027#-ND1Raw79eL0tCS_6Z9x:-ND2Q_un6p8eLRWIIM_t:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/3e456575be87ecc61a0b7677e1c7419fd936fc7a/100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java#L56)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Collections.singleton(clusterConfig.getTag()) instead?
</blockquote></details>

Done.
___
*[`100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java` line 59 at r2](https://reviewable.io/reviews/harness/harness-core/38027#-ND12FGftubx7xo7taG_-r2-59:-ND2QbTWBYwjGAq9SzIL:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/3e456575be87ecc61a0b7677e1c7419fd936fc7a/100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java#L59)):*
<details><summary><i>Previously, puthrayaharness (Deepak Puthraya) wrote…</i></summary><blockquote>

Please create a separate function for each of the modes inside AWSConnectorImpl. Then just call the function here. The code will be more readable that way
</blockquote></details>

Done.
___
*[`100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java` line 65 at r2](https://reviewable.io/reviews/harness/harness-core/38027#-ND1QfxrDx7WoQwyhFgq:-ND2QcBD6qt0x_jKh4cB:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/3e456575be87ecc61a0b7677e1c7419fd936fc7a/100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java#L65)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

We need the same CrossAccountAccessDTO for every mode. Can we build it outside and use it?
</blockquote></details>

Done.
___
*[`100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java` line 90 at r2](https://reviewable.io/reviews/harness/harness-core/38027#-ND12FGftubx7xo7taG_-r2-90:-ND2Qcu_-zjpI6wXZlK7:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/3e456575be87ecc61a0b7677e1c7419fd936fc7a/100-migrator/src/main/java/io/harness/ngmigration/connector/AWSConnectorImpl.java#L90)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

@goelpratyush99 We could use this in Azure and HTTP helm as well.
</blockquote></details>

Done.
___
*[`100-migrator/src/main/java/io/harness/ngmigration/service/entity/ConnectorMigrationService.java` line 123 at r2](https://reviewable.io/reviews/harness/harness-core/38027#-ND12FGftubx7xo7taGb-r2-123:-ND2QgFF5yVHZm89OACa:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/3e456575be87ecc61a0b7677e1c7419fd936fc7a/100-migrator/src/main/java/io/harness/ngmigration/service/entity/ConnectorMigrationService.java#L123)):*
<details><summary><i>Previously, puthrayaharness (Deepak Puthraya) wrote…</i></summary><blockquote>

We should not be doing this. If the connector has multiple secrets then we need to support fetching multiple secrets. 
</blockquote></details>

Done.

**Review by vikyathharekal — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38027)* status: 3 file(s) to review. 1 change requests. 3. unresolved discussions.

**Review by puthrayaharness — APPROVED** [review]

**Pull request merged**